Register or QuantumCircuit raises when an ambiguous non-integer gets passed to

albi3ro opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Fixes Issue #4855 .

Running any of:
* `QuantumCircuit(3.7,3.5)`
* `QuantumRegister(4.2)`
* `ClassicalRegister(5.6)`

will raise a warning.

I use <b>warn</b> instead of <b>QiskitError</b> as `float(3)` should still be valid.

I have also added tests to check that warnings are raised.

1ucian0 commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -140,6 +140,9 @@ class QuantumCircuit:
     def __init__(self, *regs, name=None, global_phase=0):
         if any([not isinstance(reg, (QuantumRegister, ClassicalRegister)) for reg in regs]):
             try:
+                if any([not isinstance(reg, (int, np.int, np.int32, np.int64)) for reg in regs]):
```
I'm not sure about that list of types. What about something like this?
```suggestion
                if any([int(reg) != reg for reg in regs]):
```
and raise an exception instead of a warning.

albi3ro replied:
Thanks for the feedback.  That would be better. 

Though, what would you think of using `math.isclose(int(reg), reg)` to handle numbers that have picked up small floating point errors?

1ucian0 commented on test/python/circuit/test_circuit_properties.py:
```diff
@@ -59,6 +61,15 @@ def test_qarg_numpy_int_exception(self):
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always")
+
+            QuantumCircuit(3.5, 2.2)
```
I think something like `QuantumCircuit(3.0)` should be accepted.

1ucian0 commented on test/python/circuit/test_circuit_registers.py:
```diff
@@ -61,6 +62,16 @@ def test_qarg_string_size(self):
+    def test_qarg_noninteger_float(self):
+        """ Test warning when passing non-integer float to QuantumRegister.
+        """
+        with warnings.catch_warnings(record=True) as w:
```
In many other tests we use [`assertWarns`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertWarns). Is there any advantage in using this style over that one?

1ucian0 commented on qiskit/circuit/library/fourier_checking.py:
```diff
@@ -84,7 +84,7 @@ def __init__(self,
                                "{1, -1}.")
 
-        super().__init__(num_qubits, name="fc: %s, %s" % (f, g))
+        super().__init__(int(num_qubits), name="fc: %s, %s" % (f, g))
```
This will be casted in the `QuantumCircuit` already. Is there any scenario where is better to capture this here?

albi3ro commented:
Thank you for your feedback.  

I now use `np.isclose` to give the inputs some more flexibility, and raise a `CircuitError` instead of just a warning.

Cryoris commented on test/python/circuit/test_circuit_properties.py:
```diff
@@ -59,6 +60,13 @@ def test_qarg_numpy_int_exception(self):
+        """
+        self.assertRaises(CircuitError, QuantumCircuit, 2.2)
+        # but an integer float should pass
+        QuantumCircuit(2.0)
```
Can you add a simple test here (also to the other one below)? Something like
```
qc = QuantumCircuit(2.0)
self.assertEqual(qc.num_qubits, 2)
```

albi3ro replied:
I have now added this to testing both `QuantumCircuit` and `QuantumRegister`.

Cryoris commented on qiskit/circuit/register.py:
```diff
@@ -41,10 +41,17 @@ def __init__(self, size, name=None):
                                % (type(size).__name__, size))
+
+        if not np.isclose(size_int, size):
+            raise CircuitError("Register size must be approximately an integer. " +
```
To be more precise we should say it must be positive
```suggestion
            raise CircuitError("Register size must be approximately a positive integer. " +
```

albi3ro replied:
Changed.

Though that if-statement is not checking to see if its positive.  By specifying positive here, it may confuse exactly why we are throwing the error.

albi3ro commented:
I have updated based on @Cryoris  's feedback.  

I also changed `int(n)` to `int(round(n))` so that a number like 1.999999 will be accepted.  This makes the code more tolerant to slight floating point errors.

Cryoris commented:
Thanks @albi3ro! 

> I also changed int(n) to int(round(n)) so that a number like 1.999999 will be accepted. This makes the code more tolerant to slight floating point errors.

I personally think we shouldn't support this. Accepting floats that are obviously integers (like `3.0`) is already pretty flexible. Register sizes are integer numbers, so it should be fine to require users to round the numbers themselves, if the register size is based on a computation.

Cryoris commented on releasenotes/notes/must_pass_integer_num_bits-98d509c2ccfc53da.yaml:
```diff
@@ -0,0 +1,7 @@
+  - |
+    The objects :class:`~qiskit.circuit.Register` and 
+    :class:`~qiskit.circuit.QuantumCircuit` now require the number of qubits
+    to be approximately (to 1e-5) integers. Non-integer floats now raise
```
```suggestion
    to be approximately integers (including e.g. 2.0). Non-integer floats now raise
```

1ucian0 commented on releasenotes/notes/must_pass_integer_num_bits-98d509c2ccfc53da.yaml:
```diff
@@ -0,0 +1,7 @@
+  - |
+    The objects :class:`~qiskit.circuit.Register` and 
+    :class:`~qiskit.circuit.QuantumCircuit` now require the number of qubits
+    to be approximately integers (including e.g. 2.0). Non-integer floats now raise
```
```suggestion
    to be integers (or castable to unambiguous integers, e.g. 2.0). Non-integer now raise
```

1ucian0 commented on qiskit/circuit/register.py:
```diff
@@ -38,11 +38,11 @@ def __init__(self, size, name=None):
-            raise CircuitError("Register size must be castable to an int (%s '%s' was provided)"
-                               % (type(size).__name__, size))
+        if int(size) != size:
+            raise ValueError("Register size must be an integer. (%s '%s' was provided)"
```
It should be `CircuitError`

1ucian0 reviewed (dismissed):
Thanks!
